[Merged by Bors] - Add "Changelog" and "Migration Guide" to PR template

MrGVSV opened the pull request:
# Objective

Context: [Discord Discussion](https://discord.com/channels/691052431525675048/745355529777315850/950532143325519902)

Improve the PR template by adding "Changelog" and "Migration Guide" sections. These sections should hopefully help speed up the review/merge process, as well as help make release notes and migration guides.

## Solution

Added the "Changelog" section template which suggests listing out the changes of the PR. This also acts as a sort of tl;dr for reviewers (especially for larger PRs).

Added the "Migration Guide" section template which suggests describing how a user might need to migrate their codebase to account for the changes by the PR. This also helps authors/contributors keep the end-user in mind when adding or changing the API.

Both sections are optional— an author does not _need_ to fill these out. Hopefully they will, though, as it provides a handful of really great benefits.

alice-i-cecile commented on .github/pull_request_template.md:
```diff
@@ -6,3 +6,12 @@
+- List a general overview of the changes made by this PR
+- Try to be concise. If more detail is needed for a particular change, consider adding it to the "Solution" section
+
+## Migration Guide
```
```suggestion
## \[Optional\] Migration Guide
```

MrGVSV replied:
Should this be added to Changelog as well?

alice-i-cecile commented on .github/pull_request_template.md:
```diff
@@ -6,3 +6,12 @@
+
+## Migration Guide
+
+- If this PR affects the public API, describe how a user might need to migrate their code to support these changes
```
```suggestion
- If this PR is a breaking change (relative to the last release of Bevy), describe how a user might need to migrate their code to support these changes
- Simply adding new functionality is not a breaking change.
- Fixing behavior that was definitely a bug, rather than a questionable design choice is not a breaking change.
```

IceSentry replied:
How can a breaking change be relative to anything other than the last release of bevy? A breaking change to something that is only in main?

MrGVSV replied:
> How can a breaking change be relative to anything other than the last release of bevy? A breaking change to something that is only in main?

That's a good point, but I think some people might see this as relative to main (especially if they're actively using main). It's probably good to specify so there's no confusion.

MrGVSV commented on .github/pull_request_template.md:
```diff
@@ -6,3 +6,12 @@
 
 - Describe the solution used to achieve the objective above.
+
+## Changelog
```
Should this include the `### Added`, `### Changed`, etc. headers?

IceSentry replied:
This feels like a lot of boilerplate for small PRs. I would maybe only add a comment suggesting using those sections

MrGVSV replied:
```markdown
- What changed as a result of this PR?
- If applicable, organize changes under "Added", "Changed", or "Fixed" sub-headings
```

Thoughts?

IceSentry replied:
LGTM

alice-i-cecile reviewed (changes requested):
IMO Changelog and Migration Guide belong at the top of the template: they're the most important to be able to skim, and the changelog provides very helpful context for reviewer.

MrGVSV commented on .github/pull_request_template.md:
```diff
@@ -6,3 +6,12 @@
+## Changelog
+
+- List a general overview of the changes made by this PR
+- Try to be concise. If more detail is needed for a particular change, consider adding it to the "Solution" section
```
Is this worth specifying?

alice-i-cecile replied:
```suggestion
- Stick to one or two sentences. If more detail is needed for a particular change, consider adding it to the "Solution" section
  - If you can't summarize the work your change may be unreasonably large / unrelated. Consider splitting your PR to make it easier to review and merge!
```

alice-i-cecile commented on .github/pull_request_template.md:
```diff
@@ -6,3 +6,12 @@
+
+## Changelog
+
+- List a general overview of the changes made by this PR
```
```suggestion
- What changed as a result of this PR?
```

alice-i-cecile commented on .github/pull_request_template.md:
```diff
@@ -6,3 +6,12 @@
 
 - Describe the solution used to achieve the objective above.
+
+## Changelog
```
```suggestion
## \[ Optional \] Changelog
```

IMO this should be optional: trivial bug fixes, doc improvements and so on don't need anything beyond Problem and Solution.

alice-i-cecile replied:
If we take this route, we should explain *why* it's optional in the template.

MrGVSV commented:
> IMO Changelog and Migration Guide belong at the top of the template: they're the most important to be able to skim, and the changelog provides very helpful context for reviewer.

Should they really go above the objective of the PR? Would this affect how authors view the `# Objective` section?

alice-i-cecile commented:
> Should they really go above the objective of the PR? Would this affect how authors view the # Objective section?

IMO the point of "Objective" is two-fold. First, to link any issues that are closed. Secondly, to help provide stronger context on *why* the problem is worth solving. These are both valuable, but feel like they're less essential to a casual reader (especially post-merge when browsing changes in release notes) than "here's what changed".

I don't feel super strongly about this though 😅

MrGVSV commented:
Another consideration are smaller PRs. If a PR fixes something small like a typo or something similar, do we want to suggest a preference for filling out the template? Such as only filling out the Objective section, or only the Changelog section.

Otherwise we might get:

```
## Objective 

Fixed typo.

...

## Changelog

### Changed

* Fixed typo.
```

(Which isn't inherently bad haha just unnecessary)

alice-i-cecile commented on .github/pull_request_template.md:
```diff
@@ -6,3 +6,15 @@
+
+## \[ Optional \] Changelog
+
+- What changed as a result of this PR?
```
```suggestion
- What changed as a result of this PR? If this was a trivial fix, or has no externally-visible impact, feel free to skip this section.
```

MrGVSV commented:
> > Should they really go above the objective of the PR? Would this affect how authors view the # Objective section?
> 
> IMO the point of "Objective" is two-fold. First, to link any issues that are closed. Secondly, to help provide stronger context on _why_ the problem is worth solving. These are both valuable, but feel like they're less essential to a casual reader (especially post-merge when browsing changes in release notes) than "here's what changed".
> 
> I don't feel super strongly about this though 😅

Maybe we add a divider between the two sections and the rest of the template? To further indicate that these are two distinct parts of the PR? And the "real" PR starts in the lower section (objective, solution, etc.)

alice-i-cecile commented:
>Another consideration are smaller PRs. If a PR fixes something small like a typo or something similar, do we want to suggest a preference for filling out the template? Such as only filling out the Objective section, or only the Changelog section.

Hmm. I've seen this problem occur even now, where the problem is trivial, as is the solution. Maybe it's Problem and Solution that should be optional?

MrGVSV commented:
> > Another consideration are smaller PRs. If a PR fixes something small like a typo or something similar, do we want to suggest a preference for filling out the template? Such as only filling out the Objective section, or only the Changelog section.
> 
> Hmm. I've seen this problem occur even now, where the problem is trivial, as is the solution. Maybe it's Problem and Solution that should be optional?

Well we probably don't want them optional for larger (or even medium-sized) PRs.

IceSentry commented:
For the migration guide I think it should be at the bottom. I don't think it helps that much at getting a quick overview.

For the changelog I'm less sure. It's nice to get a quick overview, but a changelog without the full context of the objective isn't necessarily better.

MrGVSV commented on .github/pull_request_template.md:
```diff
@@ -6,3 +6,18 @@
 
 - Describe the solution used to achieve the objective above.
+
+---
+
+## \[ Optional \] Changelog
```
@alice-i-cecile @IceSentry So should we move the changelog and migration guide to the top or keep it at the bottom?

IceSentry replied:
Personally, I prefer both at the bottom, but I'm not opposed to either approach.

alice-i-cecile replied:
I think we should probably do both at the bottom for consistency then.

MrGVSV replied:
Yeah, I think keeping them at the bottom makes sense.

> It's nice to get a quick overview, but a changelog without the full context of the objective isn't necessarily better.

I think this is a good point. It's good at top for understanding a PR at a glance, but might be more confusing to an author conceptually (i.e. it sorta disrupts their mental flow).

MrGVSV replied:
I'll leave it at the bottom then, unless anyone else has objections

MrGVSV commented on .github/pull_request_template.md:
```diff
@@ -6,3 +6,18 @@
+
+---
+
+## \[Optional\] Changelog
+
+- What changed as a result of this PR? If this was a trivial fix, or has no externally-visible impact, feel free to skip this section.
```
Thoughts on something like this:

```suggestion
## Changelog

> This section is optional. If this was a trivial fix, or has no externally-visible impact, feel free to skip this section.

- What changed as a result of this PR?
```

Just to make it easier to highlight and delete all contents, so you don't have to manually remove the `[Optional]` part.

MrGVSV replied:
@alice-i-cecile Sorry for the post-ready-for-final-review suggestion 😅

alice-i-cecile replied:
> Just to make it easier to highlight and delete all contents, so you don't have to manually remove the [Optional] part.

Yeah, okay, this is a nicer workflow, and perfectly clear still.

MrGVSV commented on .github/pull_request_template.md:
```diff
@@ -6,3 +6,18 @@
+- Stick to one or two sentences. If more detail is needed for a particular change, consider adding it to the "Solution" section
+  - If you can't summarize the work, your change may be unreasonably large / unrelated. Consider splitting your PR to make it easier to review and merge!
+
+## \[Optional\] Migration Guide
+
```
And this:

```suggestion
## Migration Guide

> This section is optional

```

cart commented:
bors r+
